content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst

```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
+not appear in the source code. The developer has let the compiler do
+that work. An additional benefit of this approach is that the compiler
+will change the shifting and masking code for us if we change the
+explicit type declarations.
+
+Why is simplicity so important? Simplicity directly increases
+understandability, which directly affects correctness and
+maintainability, which greatly affects the economic cost of the
+software. In large, long-lived projects, maintenance is by far the
+largest economic cost driver. In high-integrity applications,
+correctness is essential. Therefore, doing anything reasonable to keep
+the code as simple as possible is usually worth the effort. In some
+projects the non-functional requirements, especially performance, can
+dictate less simple code, but that won't apply to all of the code. Where
+possible, simplicity rules.
+
+One more point about the GPIO ports. There are as many of these ports as
+the Arm microcontroller vendor decides to implement. And as we said,
+they are memory-mapped, at addresses specified by the vendor. If the
+memory used by all the ports is contiguous, we can conveniently use an
+array of the :ada:`GPIO_Port` record type to represent all the ports
+implemented. We would just set the array object's address at the address
+specified for the first port object in memory. Then, normal array
+indexing will provide access to any given port in the memory-mapped
+hardware. This array approach requires each array component |mdash| the
+:ada:`GPIO_Port` record type |mdash| to be the right size so that all
+the array components start on addresses corresponding to the start of
+the next port in hardware.
+
+That starting address correspondence for the array components is
+obtained automatically as long as the record type includes all the
+memory used by any individual device. In that case the next array
+component will indeed start at an address matching the next device in
+hardware. Note that this assumes the first array component matches the
+address of the first hardware device in memory. The first array
+component is at the same address as the whole array object itself (a
+fact that is guaranteed by the language), so the array address must be
+set to whatever the vendor documentation specified for the first port.
+
+However, in some cases the vendor will leave gaps of unused memory for
+complicated memory-mapped objects like the ports. They do so for the
+sake of future expansion of the implementation, e.g., to add new
+features or capacity. The gaps are thus between consecutive hardware
+devices.
+
+These gaps are not necessarily included in the memory layout documented
+for the device. You should check, therefore, that the documented
+starting addresses of the second and subsequent array components are
+what you will get with a simple array object having components of that
+record type.
+
+To represent the gap, an "extra", unused record component can be added,
```

Comment:
A simple / short code example would be nice.

Reply:
Will do